app/src/main/java/ch/hikemate/app/model/route/HikesViewModel.kt

```diff
@@ -0,0 +1,1166 @@
+import kotlinx.coroutines.withContext
+import org.osmdroid.util.BoundingBox
+
+/**
+ * View model to work with hikes.
+ *
+ * This view model is responsible for a central part of the app's functionality, it handles hikes.
+ *
+ * @param savedHikesRepo The repository to work with saved hikes.
+ * @param osmHikesRepo The repository to work with hikes from OpenStreetMap.
+ * @param elevationRepo The service to retrieve elevation data.
+ * @param dispatcher The dispatcher to be used to launch coroutines.
+ */
+class HikesViewModel(
+    private val savedHikesRepo: SavedHikesRepository,
+    private val osmHikesRepo: HikeRoutesRepository,
+    private val elevationRepo: ElevationService,
+    private val dispatcher: CoroutineDispatcher = Dispatchers.IO
+) : ViewModel() {
+  companion object {
+    private const val LOG_TAG = "HikesViewModel"
+  }
+
+  private val _hikesMutex = Mutex()
+
+  private val _savedHikesMap = mutableMapOf<String, SavedHike>()
+
+  private var _hikeFlowsMap = mutableMapOf<String, MutableStateFlow<Hike>>()
+
+  private val _loading = MutableStateFlow(false)
+
+  private val _hikeFlowsList = MutableStateFlow<List<StateFlow<Hike>>>(emptyList())
+
+  private fun updateHikeFlowsList() {
+    _hikeFlowsList.value = _hikeFlowsMap.values.toList()
+  }
+
+  private var _selectedHikeId: String? = null
+  private val _selectedHike = MutableStateFlow<Hike?>(null)
+
+  /**
+   * Helper function to update the selected hike once (or before) [hikeFlows] has been updated.
+   *
+   * This function does not acquire the [_hikesMutex]. It is the responsibility of the caller to
+   * call this function inside of a [Mutex.withLock] block.
+   *
+   * This function does not switch context either, it is the responsibility of the caller to call
+   * this function inside of a [withContext] block.
+   */
+  private fun updateSelectedHike() {
+    // If there is no selected hike, don't bother.
+    if (_selectedHikeId == null) {
+      return
+    }
+
+    // Retrieve the selected hike from the map. If it's null, don't bother.
+    val selectedHike = _selectedHike.value ?: return
+
+    // Retrieve the corresponding flow from the map.
+    val selectedHikeFlow = _hikeFlowsMap[_selectedHikeId]
+
+    if (selectedHikeFlow == null) {
+      // The selected hike is not in the map, unselect it.
+      _selectedHikeId = null
+      _selectedHike.value = null
+    } else {
+      // The selected hike is still in the map, update it.
+      val flowValue = selectedHikeFlow.value
+      if (flowValue != selectedHike) {
+        _selectedHike.value = flowValue
+      }
+    }
+  }
+
+  /** Internal type used to store where the hikes in [_hikeFlowsMap] come from. */
+  private enum class LoadedHikes {
+    /** No hikes have been loaded yet. */
+    None,
+
+    /** The hikes loaded in [_hikeFlowsMap] are the saved hikes. */
+    FromSaved,
+
+    /** The hikes loaded in [_hikeFlowsMap] were fetched from within a geographical rectangle. */
+    FromBounds
+  }
```

Comment:
I had to check online, but no, in Kotlin a variable of type `MyEnum` can only hold *one single* value of the ones available in the enum at any given time, as opposed to languages like C# where what an enum really is is a mask of `Int`s.

Anyway, this means the hikes can either be loaded as saved hikes or from bounds, not both at the same time.

Reply:
My question was more about the logic of the code rather than the kotlin behavior. This ViewModel only provides one or the other at a single time is it?

Reply:
Okay! That's what I wanted to be sure! Thanks!